File: I2C/src/16/16.c
// Calculate pull-up resistor values

#include <stdio.h>
#include <math.h>

// I2C speed modes
typedef enum {
    I2C_STANDARD_MODE = 0,    // 100 kHz
    I2C_FAST_MODE = 1,        // 400 kHz
    I2C_FAST_MODE_PLUS = 2,   // 1 MHz
    I2C_HIGH_SPEED_MODE = 3   // 3.4 MHz
} i2c_speed_mode_t;

// Configuration structure
typedef struct {
    float vcc;                 // Supply voltage (V)
    float vol_max;            // Max LOW-level output voltage (V)
    float iol;                // LOW-level output current (A)
    float bus_capacitance;    // Total bus capacitance (pF)
    i2c_speed_mode_t mode;    // Speed mode
} i2c_config_t;

// Result structure
typedef struct {
    float min_resistance;     // Minimum resistor value (ohms)
    float max_resistance;     // Maximum resistor value (ohms)
    float recommended;        // Recommended value (ohms)
    int is_valid;            // 1 if valid range exists
} resistor_result_t;

// Get maximum rise time for speed mode (nanoseconds)
float get_max_rise_time(i2c_speed_mode_t mode) {
    switch (mode) {
        case I2C_STANDARD_MODE:
            return 1000.0;    // 1000 ns
        case I2C_FAST_MODE:
            return 300.0;     // 300 ns
        case I2C_FAST_MODE_PLUS:
            return 120.0;     // 120 ns
        case I2C_HIGH_SPEED_MODE:
            return 80.0;      // 80 ns (approximate)
        default:
            return 1000.0;
    }
}

// Get speed mode name
const char* get_speed_mode_name(i2c_speed_mode_t mode) {
    switch (mode) {
        case I2C_STANDARD_MODE:
            return "Standard Mode (100 kHz)";
        case I2C_FAST_MODE:
            return "Fast Mode (400 kHz)";
        case I2C_FAST_MODE_PLUS:
            return "Fast Mode Plus (1 MHz)";
        case I2C_HIGH_SPEED_MODE:
            return "High Speed Mode (3.4 MHz)";
        default:
            return "Unknown";
    }
}

// Calculate pull-up resistor values
resistor_result_t calculate_pullup_resistor(const i2c_config_t* config) {
    resistor_result_t result = {0};
    
    // Calculate minimum resistance
    // Rp(min) = (VCC - VOL(max)) / IOL
    result.min_resistance = (config->vcc - config->vol_max) / config->iol;
    
    // Calculate maximum resistance
    // Rp(max) = tr / (0.8473 × Cb)
    // Convert capacitance from pF to F and rise time from ns to s
    float tr_seconds = get_max_rise_time(config->mode) * 1e-9;
    float cb_farads = config->bus_capacitance * 1e-12;
    result.max_resistance = tr_seconds / (0.8473 * cb_farads);
    
    // Check if valid range exists
    result.is_valid = (result.max_resistance > result.min_resistance);
    
    // Calculate recommended value (geometric mean of min and max)
    if (result.is_valid) {
        result.recommended = sqrt(result.min_resistance * result.max_resistance);
        
        // Round to nearest standard resistor value (E12 series)
        float e12_values[] = {1.0, 1.2, 1.5, 1.8, 2.2, 2.7, 3.3, 3.9, 4.7, 5.6, 6.8, 8.2};
        float magnitude = pow(10, floor(log10(result.recommended)));
        float normalized = result.recommended / magnitude;
        
        float closest = e12_values[0];
        float min_diff = fabs(normalized - closest);
        
        for (int i = 1; i < 12; i++) {
            float diff = fabs(normalized - e12_values[i]);
            if (diff < min_diff) {
                min_diff = diff;
                closest = e12_values[i];
            }
        }
        
        result.recommended = closest * magnitude;
    }
    
    return result;
}

// Print results
void print_results(const i2c_config_t* config, const resistor_result_t* result) {
    printf("\n=== I2C Pull-up Resistor Calculation ===\n\n");
    printf("Configuration:\n");
    printf("  Supply Voltage (VCC): %.2f V\n", config->vcc);
    printf("  Bus Capacitance: %.1f pF\n", config->bus_capacitance);
    printf("  Speed Mode: %s\n", get_speed_mode_name(config->mode));
    printf("  Max Rise Time: %.0f ns\n\n", get_max_rise_time(config->mode));
    
    printf("Results:\n");
    printf("  Minimum Resistance: %.0f Ω (%.2f kΩ)\n", 
           result->min_resistance, result->min_resistance / 1000.0);
    printf("  Maximum Resistance: %.0f Ω (%.2f kΩ)\n", 
           result->max_resistance, result->max_resistance / 1000.0);
    
    if (result->is_valid) {
        printf("  Recommended Value: %.0f Ω (%.2f kΩ)\n", 
               result->recommended, result->recommended / 1000.0);
        printf("\n  ✓ Valid resistor range found!\n");
    } else {
        printf("\n  ✗ WARNING: No valid resistor range!\n");
        printf("    Bus capacitance too high or speed too fast.\n");
        printf("    Consider: reducing capacitance, lowering speed, or using active pull-ups.\n");
    }
}

int main(void) {
    // Example 1: Standard mode with typical conditions
    i2c_config_t config1 = {
        .vcc = 3.3,
        .vol_max = 0.4,
        .iol = 0.003,              // 3 mA
        .bus_capacitance = 100.0,  // 100 pF
        .mode = I2C_STANDARD_MODE
    };
    
    resistor_result_t result1 = calculate_pullup_resistor(&config1);
    print_results(&config1, &result1);
    
    // Example 2: Fast mode with higher capacitance
    printf("\n\n");
    i2c_config_t config2 = {
        .vcc = 5.0,
        .vol_max = 0.4,
        .iol = 0.003,
        .bus_capacitance = 300.0,  // 300 pF
        .mode = I2C_FAST_MODE
    };
    
    resistor_result_t result2 = calculate_pullup_resistor(&config2);
    print_results(&config2, &result2);
    
    // Example 3: Problematic case - too much capacitance
    printf("\n\n");
    i2c_config_t config3 = {
        .vcc = 3.3,
        .vol_max = 0.4,
        .iol = 0.003,
        .bus_capacitance = 500.0,  // High capacitance
        .mode = I2C_FAST_MODE_PLUS
    };
    
    resistor_result_t result3 = calculate_pullup_resistor(&config3);
    print_results(&config3, &result3);
    
    return 0;
}

File: I2C/src/16/16.rs
/// I2C Pull-up Resistor Calculator in Rust
/// Provides type-safe calculation with comprehensive validation

use std::fmt;

/// I2C speed modes with associated timing requirements
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum I2cSpeedMode {
    Standard,    // 100 kHz
    Fast,        // 400 kHz
    FastPlus,    // 1 MHz
    HighSpeed,   // 3.4 MHz
}

impl I2cSpeedMode {
    /// Get maximum rise time in nanoseconds
    pub fn max_rise_time_ns(&self) -> f64 {
        match self {
            I2cSpeedMode::Standard => 1000.0,
            I2cSpeedMode::Fast => 300.0,
            I2cSpeedMode::FastPlus => 120.0,
            I2cSpeedMode::HighSpeed => 80.0,
        }
    }
    
    /// Get frequency in Hz
    pub fn frequency_hz(&self) -> u32 {
        match self {
            I2cSpeedMode::Standard => 100_000,
            I2cSpeedMode::Fast => 400_000,
            I2cSpeedMode::FastPlus => 1_000_000,
            I2cSpeedMode::HighSpeed => 3_400_000,
        }
    }
    
    /// Get maximum bus capacitance in picofarads
    pub fn max_capacitance_pf(&self) -> f64 {
        match self {
            I2cSpeedMode::Standard => 400.0,
            I2cSpeedMode::Fast => 400.0,
            I2cSpeedMode::FastPlus => 550.0,
            I2cSpeedMode::HighSpeed => 100.0,
        }
    }
}

impl fmt::Display for I2cSpeedMode {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            I2cSpeedMode::Standard => write!(f, "Standard Mode (100 kHz)"),
            I2cSpeedMode::Fast => write!(f, "Fast Mode (400 kHz)"),
            I2cSpeedMode::FastPlus => write!(f, "Fast Mode Plus (1 MHz)"),
            I2cSpeedMode::HighSpeed => write!(f, "High Speed Mode (3.4 MHz)"),
        }
    }
}

/// Configuration for pull-up resistor calculation
#[derive(Debug, Clone)]
pub struct I2cConfig {
    pub vcc: f64,              // Supply voltage (V)
    pub vol_max: f64,          // Max LOW output voltage (V)
    pub iol: f64,              // LOW output current (A)
    pub bus_capacitance: f64,  // Total capacitance (pF)
    pub speed_mode: I2cSpeedMode,
}

impl Default for I2cConfig {
    fn default() -> Self {
        I2cConfig {
            vcc: 3.3,
            vol_max: 0.4,
            iol: 0.003,  // 3 mA
            bus_capacitance: 100.0,
            speed_mode: I2cSpeedMode::Standard,
        }
    }
}

/// Result of pull-up resistor calculation
#[derive(Debug, Clone)]
pub struct ResistorResult {
    pub min_resistance: f64,     // Ohms
    pub max_resistance: f64,     // Ohms
    pub recommended: Option<f64>, // Ohms (None if invalid)
    pub warnings: Vec<String>,
}

impl ResistorResult {
    /// Check if the calculated range is valid
    pub fn is_valid(&self) -> bool {
        self.recommended.is_some()
    }
    
    /// Get recommended value in kilohms
    pub fn recommended_kohms(&self) -> Option<f64> {
        self.recommended.map(|r| r / 1000.0)
    }
}

/// Pull-up resistor calculator
pub struct PullupCalculator;

impl PullupCalculator {
    /// Calculate pull-up resistor values
    pub fn calculate(config: &I2cConfig) -> ResistorResult {
        let mut warnings = Vec::new();
        
        // Validate configuration
        if config.bus_capacitance > config.speed_mode.max_capacitance_pf() {
            warnings.push(format!(
                "Bus capacitance ({:.1} pF) exceeds maximum for {} ({:.1} pF)",
                config.bus_capacitance,
                config.speed_mode,
                config.speed_mode.max_capacitance_pf()
            ));
        }
        
        // Calculate minimum resistance: (VCC - VOL_max) / IOL
        let min_resistance = (config.vcc - config.vol_max) / config.iol;
        
        // Calculate maximum resistance: tr / (0.8473 × Cb)
        let tr_seconds = config.speed_mode.max_rise_time_ns() * 1e-9;
        let cb_farads = config.bus_capacitance * 1e-12;
        let max_resistance = tr_seconds / (0.8473 * cb_farads);
        
        // Determine if valid range exists and calculate recommendation
        let recommended = if max_resistance > min_resistance {
            // Geometric mean for optimal performance
            let geometric_mean = (min_resistance * max_resistance).sqrt();
            Some(Self::round_to_standard_value(geometric_mean))
        } else {
            warnings.push(
                "No valid resistor range! Consider: reducing capacitance, \
                 lowering speed, or using active pull-ups.".to_string()
            );
            None
        };
        
        ResistorResult {
            min_resistance,
            max_resistance,
            recommended,
            warnings,
        }
    }
    
    /// Round to nearest E12 standard resistor value
    fn round_to_standard_value(resistance: f64) -> f64 {
        const E12_VALUES: [f64; 12] = [
            1.0, 1.2, 1.5, 1.8, 2.2, 2.7, 3.3, 3.9, 4.7, 5.6, 6.8, 8.2
        ];
        
        let magnitude = 10_f64.powf(resistance.log10().floor());
        let normalized = resistance / magnitude;
        
        let closest = E12_VALUES
            .iter()
            .min_by(|a, b| {
                let diff_a = (normalized - **a).abs();
                let diff_b = (normalized - **b).abs();
                diff_a.partial_cmp(&diff_b).unwrap()
            })
            .unwrap();
        
        closest * magnitude
    }
    
    /// Estimate bus capacitance from physical parameters
    pub fn estimate_capacitance(
        num_devices: usize,
        wire_length_cm: f64,
        include_pcb_parasitic: bool,
    ) -> f64 {
        let device_cap = num_devices as f64 * 7.5; // ~7.5 pF per device
        let wire_cap = wire_length_cm * 1.5;        // ~1.5 pF per cm
        let pcb_cap = if include_pcb_parasitic { 20.0 } else { 0.0 };
        
        device_cap + wire_cap + pcb_cap
    }
}

/// Pretty print results
impl fmt::Display for ResistorResult {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Pull-up Resistor Calculation Results:")?;
        writeln!(f, "  Minimum: {:.0} Ω ({:.2} kΩ)", 
                 self.min_resistance, self.min_resistance / 1000.0)?;
        writeln!(f, "  Maximum: {:.0} Ω ({:.2} kΩ)", 
                 self.max_resistance, self.max_resistance / 1000.0)?;
        
        if let Some(rec) = self.recommended {
            writeln!(f, "  Recommended: {:.0} Ω ({:.2} kΩ)", rec, rec / 1000.0)?;
            writeln!(f, "  Status: ✓ Valid range")?;
        } else {
            writeln!(f, "  Status: ✗ No valid range")?;
        }
        
        if !self.warnings.is_empty() {
            writeln!(f, "\nWarnings:")?;
            for warning in &self.warnings {
                writeln!(f, "  • {}", warning)?;
            }
        }
        
        Ok(())
    }
}

fn main() {
    println!("=== I2C Pull-up Resistor Calculator ===\n");
    
    // Example 1: Standard configuration
    println!("Example 1: Typical 3.3V system");
    let config1 = I2cConfig {
        vcc: 3.3,
        bus_capacitance: PullupCalculator::estimate_capacitance(2, 10.0, true),
        speed_mode: I2cSpeedMode::Standard,
        ..Default::default()
    };
    
    let result1 = PullupCalculator::calculate(&config1);
    println!("{}", result1);
    
    // Example 2: Fast mode with higher capacitance
    println!("\nExample 2: Fast Mode with multiple devices");
    let config2 = I2cConfig {
        vcc: 5.0,
        bus_capacitance: PullupCalculator::estimate_capacitance(4, 20.0, true),
        speed_mode: I2cSpeedMode::Fast,
        ..Default::default()
    };
    
    let result2 = PullupCalculator::calculate(&config2);
    println!("{}", result2);
    
    // Example 3: Edge case - too much capacitance
    println!("\nExample 3: Problematic configuration (high capacitance)");
    let config3 = I2cConfig {
        vcc: 3.3,
        bus_capacitance: 500.0,
        speed_mode: I2cSpeedMode::FastPlus,
        ..Default::default()
    };
    
    let result3 = PullupCalculator::calculate(&config3);
    println!("{}", result3);
    
    // Example 4: Using type safety
    println!("\nExample 4: High-speed mode requirements");
    for mode in [I2cSpeedMode::Standard, I2cSpeedMode::Fast, 
                 I2cSpeedMode::FastPlus, I2cSpeedMode::HighSpeed] {
        println!("\n{}:", mode);
        println!("  Max rise time: {:.0} ns", mode.max_rise_time_ns());
        println!("  Max capacitance: {:.0} pF", mode.max_capacitance_pf());
        println!("  Frequency: {} kHz", mode.frequency_hz() / 1000);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_standard_mode_calculation() {
        let config = I2cConfig::default();
        let result = PullupCalculator::calculate(&config);
        
        assert!(result.is_valid());
        assert!(result.recommended.unwrap() > result.min_resistance);
        assert!(result.recommended.unwrap() < result.max_resistance);
    }
    
    #[test]
    fn test_invalid_configuration() {
        let config = I2cConfig {
            bus_capacitance: 1000.0, // Very high capacitance
            speed_mode: I2cSpeedMode::HighSpeed,
            ..Default::default()
        };
        
        let result = PullupCalculator::calculate(&config);
        assert!(!result.is_valid());
        assert!(!result.warnings.is_empty());
    }
    
    #[test]
    fn test_capacitance_estimation() {
        let cap = PullupCalculator::estimate_capacitance(3, 15.0, true);
        assert!(cap > 0.0);
        // 3 devices * 7.5 + 15cm * 1.5 + 20 = 64.5 pF
        assert!((cap - 64.5).abs() < 0.1);
    }
}